src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/X509Certificates/X509CertificateKeyAccessors.cs

```diff
@@ -221,5 +370,411 @@ private static ArraySegment<byte> GetCertificateSubjectPublicKeyInfo(X509Certifi
+
+                if (clone != null)
+                {
+                    return clone;
+                }
+            }
+
+            if (privateKey is MLDsaImplementation mldsaImplementation)
+            {
+                using (CngKey clonedKey = mldsaImplementation.CreateEphemeralCng())
+                {
+                    return CopyWithEphemeralKey(certificate, clonedKey);
+                }
+            }
+
+            // MLDsaCng and third-party implementations can be copied by exporting the PKCS#8 and importing it into
+            // a new MLDsaCng. An alternative to PKCS#8 would be to try the private seed and fall back to secret key,
+            // but that potentially requires two calls and wouldn't allow implementations to do anything smarter internally.
+            // Blobs may also be an option for MLDsaCng, but for now we will stick with PKCS#8.
+            byte[] exportedPkcs8 = privateKey.ExportPkcs8PrivateKey();
+
+            using (PinAndClear.Track(exportedPkcs8))
+            using (MLDsaCng clonedKey = MLDsaCng.ImportPkcs8PrivateKey(exportedPkcs8, out _))
+            {
+                CngKey clonedCngKey = clonedKey.KeyNoDuplicate;
+
+                if (clonedCngKey.AlgorithmGroup != CngAlgorithmGroup.MLDsa)
+                {
+                    Debug.Fail($"{nameof(MLDsaCng)} should only give ML-DSA keys.");
+                    throw new CryptographicException();
+                }
+
+                return CopyWithEphemeralKey(certificate, clonedCngKey);
+            }
+        }
+
+        private static SafeCertContextHandle CopyWithEphemeralKey(X509Certificate2 certificate, CngKey cngKey)
+        {
+            Debug.Assert(string.IsNullOrEmpty(cngKey.KeyName));
+
+            // Handle makes a copy of the handle.  This is required given that CertSetCertificateContextProperty accepts a SafeHandle
+            // and transfers ownership of the handle to the certificate.  We can't transfer that ownership out of the cngKey, as it's
+            // owned by the caller, so we make a copy (using Handle rather than HandleNoDuplicate).
+            using (SafeNCryptKeyHandle handle = cngKey.Handle)
+            {
+                // Make a new pal from bytes.
+                SafeCertContextHandle hCertContext = FromBlob(certificate.RawData, X509KeyStorageFlags.PersistKeySet);
+                try
+                {
+                    if (!Interop.Crypt32.CertSetCertificateContextProperty(
+                        hCertContext,
+                        Interop.Crypt32.CertContextPropId.CERT_NCRYPT_KEY_HANDLE_PROP_ID,
+                        Interop.Crypt32.CertSetPropertyFlags.CERT_SET_PROPERTY_INHIBIT_PERSIST_FLAG,
+                        handle))
+                    {
+#if NETFRAMEWORK
+                        // TODO CertSetCertificateContextProperty has SetLastError=true so I think
+                        // this should be equivalent to Marshal.GetLastPInvokeError?
```

Comment:
>  this should be equivalent to Marshal.GetLastPInvokeError?

Not exactly. The problem is sometimes the CLR itself will make a Win32 call that wipes out during p/invoke. `GetLastPInvokeError` was introduced so the runtime preserves the callers actually last p/invoke error, not what Windows says it was, because the runtime may have wiped it out.

So we use `GetLastPInvokeError` when we can.